Add some compute tags for Newtonian hydro

carmaza opened the pull request:
## Proposed changes

This PR adds some compute tags that I've been using in some turbulence simulations in Newtonian hydro, although in principle they can be used in any `NewtonianEuler` run. The main purpose of these items is to observe them in the volume, as well as observing their volume integrals for post processing operations. Since most of these items are just multiplications of simple tags, I'd be curious to know if there is a more practical way to observe them rather than adding the compute tags, in which case I'd update (or close :P) this PR accordingly. 

This code shouldn't take long to review for anyone who has run any hydro simulation. Maybe @fmahebert can provide a quick second review? I'd be interested to know if he (or any other hydro user) has better choices for the names of the files and tags that I'm introducing here.



### Types of changes:

- [ ] Bugfix
- [x ] New feature
- [ ] Refactor

### Component:

- [x ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ x] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [x ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [x ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

carmaza committed:
Add compute items for Newtonian turbulence

kidder commented on src/Evolution/Systems/NewtonianEuler/Tags.hpp:
```diff
@@ -99,6 +99,39 @@ struct VPlus : db::SimpleTag {
+
+/// The local Mach number of the flow
+template <typename DataType>
+struct Mach : db::SimpleTag {
```
I'd call this `MachNumber`

kidder commented on src/Evolution/Systems/NewtonianEuler/TurbulenceComputeItems.cpp:
```diff
@@ -0,0 +1,150 @@
+}
+
+template <typename DataType, size_t Dim, typename Fr>
+void mach(const gsl::not_null<Scalar<DataType>*> result,
```
mach_number

nilsdeppe commented on src/Evolution/Systems/NewtonianEuler/TurbulenceComputeItems.hpp:
```diff
@@ -0,0 +1,199 @@
+
+//@{
+/*!
+ * Compute the local Mach number, \f$\text{Ma} = v/c_s\f$,
```
[optional] I think the `/*!` need a `\brief` on the first line if you want them to have a brief comment, but I wasn't able to render on my machine (for unrelated reasons)

nilsdeppe reviewed (commented):
Feel free to squash things immediately

kidder commented:
I am happy for you to squash

nilsdeppe commented:
me too

carmaza commented:
Squashed and rebased on develop.

fmahebert commented on src/Evolution/Systems/NewtonianEuler/Tags.hpp:
```diff
@@ -99,6 +99,39 @@ struct VPlus : db::SimpleTag {
 // @}
 
+/// The internal energy density.
+template <typename DataType>
```
For my understanding: why are these Tags templated on DataType? Are they likely to be used for anything else than a DataVector?

carmaza replied:
I preferred to be consistent with the other tags in this file. I don't think I can ensure that these tags (or the ones already in the file) will only be used for DataVectors, but, if you prefer, I can remove that feature from all the tags in this file.

fmahebert replied:
Okay thanks for clarifying, I'd say just leave as is... I think taking a closer look at what is templated on DataType and whether it's necessary can be done separately at some later point.

fmahebert commented on tests/Unit/Evolution/Systems/NewtonianEuler/Test_TurbulenceComputeItems.cpp:
```diff
@@ -0,0 +1,177 @@
+}
+
+template <size_t Dim, typename DataType>
+void test_compute_items(const DataType& used_for_size) noexcept {
```
I'm quite confused by this test:
- it is testing SoundSpeedSquared, but that is already tested in Test_SoundSpeedSquared... the code looks copy-pasted here? Is this intentional?
- do the lines `// CHECK implementation of MachNumber here` mean there is a test you were still planning to add?
- you test that each free function (e.g., mach_number) produces the same output as the free function called through the databox (`MachNumberCompute`), but many of the free functions are not tested to have the correct result, so testing self-consistency doesn't seem useful. Could these missing correctness tests be added?

carmaza replied:
- The lines 

```
TestHelpers::db::test_compute_tag<
      NewtonianEuler::Tags::SoundSpeedSquaredCompute<DataType>>(
      "SoundSpeedSquared");
TestHelpers::db::test_compute_tag<
      NewtonianEuler::Tags::SoundSpeedCompute<DataType>>("SoundSpeed");
```

are the same as in the file you mention. At some point my intention was to fuse both files into a single one. Would that make more sense, or should I just delete the lines above if you find it overkill to leave them here too?

- You are right--in the end I opted not to test the Mach number as it's just a division. I can either add the proper test, or remove the comments.

- Line 115 tests the correct implementation of the specific kinetic energy free function. Same for line 117 and kinetic energy density, line 130 and ram pressure (expected implementation starts in line 122), line 145 and internal energy density with 1d EoS, and line 163 with 2d EoS. Additionally (and as a repeat, see the first bullet point), lines 139 and 155 check the sound speed squared implementation. Is this what seems to be missed? Also, now I'm leaning towards fusing SoundSpeedSquared.*pp to the *pp files in this PR again. Maybe also rename these files to simple ComputeItems.*pp or similar as is done in GH?

Thanks for taking a look.

fmahebert replied:
- I like the idea of merging the files (src and tests) for the SoundSpeedSquared and TurbulenceComputeItems into a single ComputeItems file, as you describe. Totally up to you if/when you'd like to merge these. But until the files are merged, I would favor removing the SoundSpeedSquared tests that are duplicated here. 
- Please test the Mach number, as simple as it may be :)
- Ah, I'd missed the tests of the energy density functions, thanks for setting me straight. I think once you add the mach_number test then everything will be tested, and the sound_speed_squared test can stay or go depending on your decision about merging.

fmahebert reviewed (changes requested):
I'm concerned that the test isn't quite complete yet

carmaza committed:
Move SoundSpeedSquared compute tag to ComputeItems file

nilsdeppe commented:
@fmahebert @kidder could you please look at the fixups?

I'm happy with you squashing once the other reviewers have looked at the fixups

fmahebert commented on src/Evolution/Systems/NewtonianEuler/ComputeItems.cpp:
```diff
@@ -100,6 +140,7 @@ Scalar<DataType> specific_kinetic_energy(
 
 #define DTYPE(data) BOOST_PP_TUPLE_ELEM(0, data)
 #define DIM(data) BOOST_PP_TUPLE_ELEM(1, data)
+#define THERMO_DIM(data) BOOST_PP_TUPLE_ELEM(1, data)
```
I think what you've done here (setting DIM, THERMO_DIM both to the the 1'th element, since you know they won't be used together) is correct, but I found I needed to read the whole instantiation section a few times to make sure I understood it.

I would favor that the types be defined and undef'd as needed around each instantiation block. It's a few more lines of code, but avoids some possible ambiguities.

fmahebert reviewed (commented):
I'm happy overall, but I found the styling of the instantiation macros confusing.. so here's a suggestion to make things easier to read.

I would be happy for you to edit this while squashing, if Larry gives his permission for you to squash

kidder commented:
I'm happy with the fixups

carmaza commented:
Squashed and rebased on develop.

@fmahebert I reordered and renamed a few things in the instantiations (I squashed right away; hope that's OK). Let me know if the new version clears up the confusion.

fmahebert reviewed (approved):
I think the new instantiations are clearer, thanks @carmaza
